### PR TITLE
Macros enabling removal of explicit Fact naming

### PR DESCRIPTION
**Pull request by jhkuperus**
Previously, `Fact`s had to be declared and named manually to guarantee uniqueness of names inside the engine:

``` scala
object MyGlossary extends Glossary {
  val MyFact = new SingularFact[Int]("MyFact")
}
```

The redundancy of the `Fact`'s name is not just a nuisance to type, it actually causes trouble when renaming the `Fact`, or trying to copy its declaration. In both situations, you _must_ also alter the name parameter manually. The compiler and IDE tooling cannot help you track these errors down and so you are bound to find out about them at runtime, with potentially uninformative error messages.

This pull request adds several macros to the project which allow for much easier declaring of `Fact`s. The macros run during compile-time and transform their invocation into the line of code above, extracting the name of the `val` and providing it as the name of the `Fact` being constructed.

Using these macros, the above example could be rewritten as follows:

``` scala
import org.scalarules.utils.FactMacros._

object MyGlossary extends Glossary {
  val MyFact = defineFact[Int]
}
```

There are alternatives of the macro for creating `SingularFact` and `ListFact` instances, as well as variations accepting a value for the `description` parameter of the base fact.

In light of this added feature, this PR also includes a bump in the minor version of the project. Backwards compatibility is guaranteed because the old method of defining `Fact`s manually is still available and unaltered.

### CODE REVIEW

**Comment by jhkuperus** [comment]
I lied a bit about the backwards compatibility. I replaced the body of the `Glossary` class and anyone currently using it will be facing compiler errors. I'm just about to push a commit restoring it to its former glory and moving the new code to a new class called `MacroGlossary`.

**Comment by jqno** [comment]
I think this is extremely cool stuff.

I'm not sure about your intent though, judging from your story. Do you mean to have the macro implementation replace the old implementation? Or do you always want to have the old one as a fallback?

I would assume the former, because of the issues you mention with renaming them etc.

But it feels like the macro implementation is still kind of a second class citizen. I notice you did deprecate the `Glossary` class, but not its methods. (Which means you can hide the warning in the imports, not in the code actually using it.) Also, the new glossary is called `MacroGlossary`, which refers to its implementation details and therefore feels uncomfortable as a general purpose solution. Are you worried that there are still issues with the macro implementation?

In other words: why not just throw out the old implementation and replace it with the macro stuff? Even in [semantic versioning](http://semver.org/), you're allowed to have breaking changes when the major version is still 0 (which it is). I'm not sure how large your client code base has become, but it doesn't seem like such a big deal to just change everything to the macro implementations there -- or am I missing something here?

But: extremely cool stuff!

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/MacroGlossary.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  def defineFact[A](description: String = "No description"): SingularFact[A] = macro FactMacros.defineFactMacroWithDescriptionImpl[A]
+
+  def defineListFact[A](): ListFact[A] = macro FactMacros.defineListFactMacroImpl[A]
+  def defineListFact[A](description: String = "No description"): ListFact[A] = macro FactMacros.defineListFactMacroWithDescriptionImpl[A]
+
+  /**
+    * Collects all declared `Fact`s in this `Glossary` and returns them mapped from their names to their definitions.
+    */
+  lazy val facts: Map[String, Fact[Any]] = {
+    val declaredFields: Array[Field] = this.getClass.getDeclaredFields
```
Comment (jqno) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/MacroGlossary.scala]: The local val looks a bit superfluous here :)

File: engine/src/test/scala/org/scalarules/utils/GlossaryTest.scala [hunk 2]
```diff
@@ -3,6 +3,7 @@ package org.scalarules.utils
 import org.scalarules.engine.SingularFact
 import org.scalatest.{FlatSpec, Matchers}
 
+@deprecated(message = "Glossary is deprecated in favor of MacroGlossary", since = "0.3.0")
```
Comment (jqno) [inline engine/src/test/scala/org/scalarules/utils/GlossaryTest.scala]: Deprecating a unit test, is that a thing? :)

Comment (jhkuperus) [inline engine/src/test/scala/org/scalarules/utils/GlossaryTest.scala]: If you have a better way of shutting the compiler up about using deprecated API... ;) 

I thought it fitting in a way, to also expressly indicate that this test is testing something that's deprecated. After all, it too will be deleted once the code under test is removed.

**Comment by jhkuperus** [comment]
As I stated out-of-bound: here I am trying to actually be backwards-compatible and you encourage me not to ;)

Anyways, you certainly have a fair point. Our user-base is probably still just 1 project, which is also in our own control. And actually the 'old way' is easy enough for someone to recreate if they're too scared of macros. 

All in favour of removing the old way without deprecation and provide the macro-way as our pre-1.0 default, please +1 this particular comment. ( @jqno @vzorge @NRBPerdijk )
(Edit: 3+ votes make it so, just to be explicit) :)
(Edit 2: Didn't await vote number 3) ;)

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/MacroGlossary.scala (hunk 1 above)
Comment (jhkuperus) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/MacroGlossary.scala]: Agreed

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala [hunk 3]
```diff
@@ -11,7 +11,7 @@ import scala.language.experimental.macros
   * Utility base class for collecting and namespacing `Fact`s. You can extend this class, define facts in it and receive a
   * utility collection of all facts declared in your class.
   */
-class MacroGlossary {
+class Glossary {
   def defineFact[A](): SingularFact[A] = macro FactMacros.defineFactMacroImpl[A]
   def defineFact[A](description: String = "No description"): SingularFact[A] = macro FactMacros.defineFactMacroWithDescriptionImpl[A]
```
Comment (vzorge) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala]: Why do we give a default for the description parameter? If none is given, I assume the no-args method is called, making this default pointless

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala [hunk 4]
```diff
@@ -6,37 +6,117 @@ import scala.annotation.compileTimeOnly
...
+    * @return a `ListFact` initialized with the name of the val declaration.
+    */
+  def defineListFactWithDescription[A](description: String): ListFact[A] = macro FactMacros.defineListFactMacroWithDescriptionImpl[A]
+
+
+  // Macro implementations
+
   @compileTimeOnly("This is a compile-time macro implementation, do not call it at runtime")
-  def defineFactMacroImpl[A : c.WeakTypeTag](c: Context)(omschrijving: c.Expr[String]): c.Expr[SingularFact[A]] =
+  def defineFactMacroImpl[A : c.WeakTypeTag](c: Context)(): c.Expr[SingularFact[A]] =
```
Comment (vzorge) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:75]: Why are these methods all called xxImpl? This feel like a reference to a (Java) interface en implementation. Perhaps a leftover name from earlier work?

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala [hunk 5]
```diff
@@ -6,37 +6,117 @@ import scala.annotation.compileTimeOnly
...
+  // Macro definitions
+
+  /**
+    * Defines a `Fact`, using the name of the `val` it is assigned to as the name of the `Fact`. Note: the value of this
+    * macro *must* be assigned to a `val`, otherwise a compiler error will be raised.
+    *
+    * @tparam A the value type of the resulting `Fact`.
+    * @return a `SingularFact` initialized with the name of the val declaration.
+    */
+  def defineFact[A](): SingularFact[A] = macro FactMacros.defineFactMacroImpl[A]
```
Comment (vzorge) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:40]: Where are these methods used? They are similar in name as the ones in the Glossary, but never used. Much better documented though than the ones in the Glossary

**Comment by vzorge** [comment]
Because of these changes, the value **false** is now always given for the _isResult_ parameter for a Fact. We should probably remove this parameter for now. Its intended use was to differentiate between types of Facts (inputs, partial results and outputs). This was always a sub-optimal solution and it would be better to redesign this from scratch

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala (hunk 3 above)
Comment (jhkuperus) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala]: In "normal" Scala-world, you would be completely correct. However, macros do not support default parameters. So, in order to support the `description` being optional, we need both these macros, one with a `description` and one without a `description` parameter.

Macro declarations need to match 1-on-1 with the signature of their macro implementation.

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala (hunk 4 above)
Comment (jhkuperus) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:75]: Nope, as far as macros go, this is a bit of a convention. Because a macro consists of a macro _definition_ and a macro _implementation_. To keep these apart, `Impl` is added as a postfix in pretty much all of the examples. Note that these `Impl`-methods are not supposed to be used on runtime. (Hence the `@compileTimeOnly` annotations).

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala (hunk 5 above)
Comment (jhkuperus) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:40]: They are used when you create your own class extending `Glossary`. Basically it comes down to this:

``` scala
class MyGlossary extends Glossary {
  val MyFact = defineFact[Int]()
}
```

**Comment by jhkuperus** [comment]
Yes, I agree. I'll remove it, since it has no value anymore whatsoever.

As a replacement, I was thinking about allowing users to define their own 'metadata' to associate with `Fact`s, but I haven't worked out a design for it yet.

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala (hunk 3 above)
Comment (vzorge) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala]: I understand the need for two methods. I was arguing that the default "no description" for the method which has a description is perhaps not neccessary. Not supplying a description will not fill in "no description", but rather call the no-arg method. And leaving it like this might cause confusion.
Unless of course I am missing some vital macro knowledge here.

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala (hunk 5 above)
Comment (vzorge) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:40]: I assumed when you extend Glossary, that you then call the methods defined in Glossary.scala. These method names are similar to the ones in FactMacros, which is why I asked my question.
I do not understand why these methods are duplicated both in Glossary.scala and FactMacros.scala

File: engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala (hunk 3 above)
Comment (jhkuperus) [inline engine/src/main/scala/org/scalarules/dsl/core/glossaries/Glossary.scala]: Ah, I see what you mean 🐹  That's actually a remnant. I tried building them with a default value, which caused me to run into the issue I described above. Now the default value is indeed useless. Removed them.

File: engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala (hunk 5 above)
Comment (jhkuperus) [inline engine-core/src/main/scala/org/scalarules/utils/FactMacros.scala:40]: Sorry, I missed the fact that we were looking at `FactMacros`. 

I initially added these to `FactMacros` to be one unit for the macros, both definitions and implementations. Here they sort of 'naturally' belong together. In some tests I use `FactMacros.defineFact` to create facts without using a `Glossary`. The `Glossary` 'duplicates' these methods so it can offer the same functionality without needing to prefix the macro calls with `FactMacros`.

We could remove the ones in `FactMacros`, since the intended use is through extending the `Glossary` class. We'll discuss it today :)

**Comment by vzorge** [comment]
What is keeping this pull request from being merged?

**Comment by jhkuperus** [comment]
Me giving @NRBPerdijk a chance to take a look at it first :)

We'll continue working on calculations again soon, I'll make sure it's merged before such time.

**Comment by NRBPerdijk** [comment]
Reviewed, looks rather nice! Macro's sure are different... :)

**Comment by jhkuperus** [comment]
Closes #11
